File: leetcode2/2medium/최원준/Q1357.java
package Leetcode.최원준;

/*
1. 아이디어 :


2. 시간복잡도 :
Cashier = O( n )
getBill = O( m )

3. 자료구조/알고리즘 :
- / -
 */

import java.util.HashMap;
import java.util.Map;

public class Q1357 {
    class Cashier {
        int n;
        int count = 0;
        int discount;
        Map<Integer, Integer> productCostMap = new HashMap<>();
        int[] productCost;

        public Cashier(int n, int discount, int[] products, int[] prices) {
            this.n = n;
            this.discount = discount;
            productCost = new int[201];

            for (int i=0; i<products.length; i++) {
                productCost[products[i]] = prices[i];
            }
        }

        public double getBill(int[] product, int[] amount) {
            double total = 0;
            for (int i=0; i<product.length; i++) {
                total += amount[i] * productCost[product[i]];
            }
            return (++count % n == 0) ? total * ((100-discount)) / 100 : total;
        }
    }

/**
 * Your Cashier object will be instantiated and called as such:
 * Cashier obj = new Cashier(n, discount, products, prices);
 * double param_1 = obj.getBill(product,amount);
 */
}

File: leetcode2/2medium/최은서/Q1357.js
/**
 * @param {number} n
 * @param {number} discount
 * @param {number[]} products
 * @param {number[]} prices
 */
var Cashier = function(n, discount, products, prices) { //생성자 함수
    this.n = n //할인을 적용하는 주기
    this.count = 0; //count 번째 고객
    this.discount= discount * 0.01 //할인율
    this.dict = {} //가격 저장
    for (let i = 0; i < products.length; i++) {
        this.dict[products[i]] = prices[i]
    }
};

/** 
 * @param {number[]} product 
 * @param {number[]} amount
 * @return {number}
 */
Cashier.prototype.getBill = function(product, amount) { //메서드
    this.count++ //현재 고객 번호 업데이트
    let res = 0;
    for (let i = 0; i < amount.length; i++) {
        res += this.dict[product[i]] * amount[i]
    } //가격 계산
    if (this.count % this.n === 0) {
        res = res - (res * this.discount)
    } // n 번째 고객이면 할인 적용 (값 - 할인율을 적용한 할인값)
    return res;
};

/** 
 * Your Cashier object will be instantiated and called as such:
 * var obj = new Cashier(n, discount, products, prices)
 * var param_1 = obj.getBill(product,amount)
 */